### PR TITLE
Separators, grouped controls, and wide controls

### LINKED ISSUE #57: Multiple buttons using the same container
Hi,

it would be nice to have lets say 6 buttons which are nested inside the same container. The container could have a title like: "Dimming" and there would be 6 buttons with different dimming levels.

It would look cleaner than having 6 individual buttons which all have the title "Dimming".

Thanks!

### LINKED ISSUE #72: Separator item (line)
I use the ESPUI as UI for my smart home devices. Depanding on the device there are a lot of items 
 - up to 20+ on some tabs - to fill arrays in my software. 
I'd like to break apart  the screen layout in logical (horizontal) blocks.

I think a separator line - can have text or not, but should be small compared to other items  may be a easy to implement and easy to use solution.

If the coaintainer idea (#57) would be available, I would use this.

(IMHO the ESPUI is absolutly great ...!)

### LINKED ISSUE #73: Different element sizes
I use the ESPUI as UI for my smart home devices. Depanding on the device there are a lots of items
up to 20+ on some tabs - to fill arrays in my software. 
Due to the (defined) element sizes, the screen is very soon filled and I need to scroll down 
Need to admit: I have no idea of web design, css... at all - so ESPUI is made for me :-)

It would be nice to have different sets of items in different sizes. They could by used like the Win10 blocks UI.
Switching between them could by implemented as a function call etc. before ESPUI.addControl. 
This could keep existing project using ESPUI compatible.

Alternative I need to learn more about CSS etc... 
This would allow to change the UI of  elements in generell (but is still not that flexible). A little help what I need to look for would be greatly appreciated !

### LINKED ISSUE #115: Is it possible to force single column on wider screen? (question)
This is actually a question, not a bug. Didn't find how to label it as question.

Is it possible to force all controls into a single column on wide screen?

The UI I created looks great on phones. On PC screen size the control order makes limited sense.

### PR DESCRIPTION
**Pull request by iangray001**
This adds some significant new layout options that quite a few people asked for. 

**This closes #57, closes #72, closes #73, closes #115, and does a good amount of what was asked for in #107 and #139.**

When considering #139, @codefaux had some good ideas, but I did not want to lose what made this library great, which was a simple no-fuss layout, all from C code, and that works on a variety of web clients. The goal therefore was to add some powerful features that allow some much-needed customisability, but without complicating anything, and without changing the default use case in any way.

Added features are **separators**, **grouped controls**, and **single column layouts** for controls. The best way to see these is to see the screenshots in [the updated documentation](https://github.com/iangray001/ESPUI/blob/designupdates/README.md).

Most of the changes are minimal. The on-device code has had a few minor parameters added but very little was required. `controls.js` on the other hand has been significantly tidied up. Previously, the HTML was copied inline dozens of times (once for each element type) with loads of repeated logic. It is now refactored into `addToHTML()` and `elementHTML()` which allows for easy updates to design elements, and crucially, separates the tasks of adding a panel and adding a control. This allows for multiple controls to be added to a single panel (and the code is much cleaner).

This PR is technically not 100% backwards compatible, but it is 99.9%. If these new features are not used then nothing changes. All existing user-facing functions are identical. The only differences that would affect an existing project are in `style.css`. Previously the tab bar had no bottom margin and cards all had a top margin. This ruins the ability to include separators near to cards, so I switched the margins to a bottom margin on the tabs and cards. Visually, this looks the same. It might vary by a pixel or two.

Also, cards used to have percentage-based right margins, which I changed to a fixed pixel amount to allow for wide elements to be sized to a predictable amount. Again this should be visually almost indistinguishable because I picked a size which looks the same.

### CODE REVIEW

**Comment by thomastech** [comment]
Nice! The Separators will really help out my UI.

I'll try out your PR this week and will report back.

- Thomas

**Comment by thomastech** [comment]
@iangray001, rather than hold things up I went ahead and tried out the new features. I added Separators, Groups, and a Wide panel. All worked, no issues.

**Future Idea / Suggestion:**
It would be great if a grouped switch could include its label (name). For example, below is a screenshot of a Text input that includes a Switch. I wanted the switch to be labeled as "Enable RDS". As a workaround I added a Label Control, but it looks tacky.
ESPUI_group

A workaround would be to apply custom CSS to the Label control so its gray box area was the same length as the label's text size and then set its background to match the Panel body color. I don't think your new Inline Styles feature can do all that. Or can it?

TLDR; Thumbs-up on the new features. Thank You!

- Thomas

**Review by s00500 — APPROVED** [review]
Hey @iangray001,
This is absolutely great, thanks a lot, really appreciate it :-)
Just one tiny note: UPDATE_ACCEL, I see you fixed the ID issue there, in general it does not do anything and can probably be removed..

Also thanks for abstracting the DOM handling, makes it much clearer to read :-)

**Comment by iangray001** [comment]
@thomastech, as you've seen, yes you can group pretty much anything although the results are unpredictable :) 

I was thinking about the labelling issue, and I agree that switchers and sliders without labels is of limited use. Again though, I didn't want to upset too much with what was already a large PR. 

I'll have a think about how to add labels to things at some point. Your workaround of using a label will work, and inline styles can indeed fix up your layout issues. I haven't got my test harness in front of me so this is from memory, but first deal with the background. You can set its background with:

```
background-color: #e74c3c;
```

(The colours used by the strange names "Alizarin" etc. are at the top of styles.css.) But if you want to just remove the grey box entirely:

```
background-color: unset;
```

You'll probably need to colour the font black:

```
color: black;
```

And lets add a little gap between the top of the label and the control it is labelling:

```
margin-top: 10px;
```

So something along these lines should tidy it up:

```
char inlinestyle[] = "background-color: unset; color: black; margin-top: 10px;";
```

**Comment by thomastech** [comment]
@iangray001, thanks for the hints on the custom CSS. Looks nice on normal size panels.

ESPUI_group1

The reason I suggested changing the Label control so its gray box area was the same length as the label's text size is to prevent layout issues on wide panels. For example:

ESPUI_group2

I'm assuming if the size was reduced it would snuggle up next to the Switch on the wide panel, but still look correct on a small panel (or so I hope). But I've looked through style.css I can't identify the tag that allows me to alter the length. Any advice is appreciated.

- Thomas

**Comment by iangray001** [comment]
Oh I see what you mean. When the card is made wide the switcher and label are allowed to flow onto the same line. I did say in the PR that grouping controls of different kinds will often have undesirable visual artefacts! :)

[Label styles are set 90% in the style.](https://github.com/s00500/ESPUI/blob/cf535110ea39718657bf7539502e967d8370a9c3/data/css/style.css#L116) Normally this forces them to be on a new line but on a wide card with narrow other elements they can fit. Just set the label to full width:

`width: 100%; background-color: unset; margin-top: 10px`

That looks quite nice to me. 

Screenshot 2022-01-10 at 09 56 29

Once I have got some more important tasks off my stack I'm going to look at huge update to the documentation and examples to show off what can be done with all this.

**Comment by thomastech** [comment]
That's what I needed. Thanks for all the hand-holding. I appreciate it.

Looking forward to your updates to the docs. BTW, it could be helpful to add a comment that setPanelStyle() and setElementStyle() must be used *AFTER* ESPUI.begin() or ESPUI.beginLITTLEFS(). That will save someone from debugging a looping boot crash if the inline styles functions are used before ESPUI has initialized.

- Thomas

**Comment by thomastech** [comment]
@iangray001, I thought I was done with all the questions. Famous last words.

I haven't run into any problems when using the custom CSS on Label controls. But the Text Input controls are a bit different. For example, I want to reduce the width of a Text input (plus increase font size). So I did something like this:
```
ESPUI.setPanelStyle(rdsStationID, "font-size: 1.5em; width: 35%;");
ESPUI.setElementStyle(rdsStationID, "width: 35%; color: black;");
```
This gives the default width (ignores 35% css), but with the larger 1.5em font. The width only changes after I submit a new value.
And another odd thing is that I had to include the _color: black_ because changing the width set the color to white.

So I need your advice on the proper way to set the width.

- Thomas

**Comment by iangray001** [comment]
> debugging a looping boot crash

Grrf, I'm annoyed I didn't notice this. Yes, @s00500 asked me to make those functions "update" functions so they must be called after starting the UI, but there is no excuse for them causing a crash. It's on the list, I'll get to that. Thanks.

> changing the width set the color to white

Ah ha, you've spotted a "quirk". For reasons that I certainly wouldn't have chosen, the text input controls are *created* with inline styles. If we go back to before I started meddling and [observe the line that adds the text input control](https://github.com/s00500/ESPUI/blob/78317ce5f2a17e45857cdef867b62cb9ee29ffa9/data/js/controls.js#L572) we can see it is created as having an inline CSS:

```
<input style='color:black;' id='text" + data.id + ...
```

Therefore if you add your own inline style it overrides this, and so yes you have to add back in the `color: black`. The correct way to do this was with an addition to the main CSS, but unfortunately there was no way for me to fix this and maintain perfect backwards compatibility, so I didn't. I presumed that once people are messing with this then they'll work it out, which you did! :)

As for your question about width, you are using the correct CSS. I suspect that the issue is related to websocket message density. There is a bug at the moment in which some messages don't get through during UI set up (or very soon after). 

If we investigate, we see that the ESP believes that it is sending all messages, but they are not making it through the WebSocket later. We can verify this by printing the value of `ws->availableForWriteAll()` before each message is sent. Eventually (in my tests after around 10 or so messages, so after 100 controls), the return value of `availableForWriteAll` becomes `false`, indicating that the internal buffers are full and so any further messages sent to the layer are simply discarded. We cannot slow transmission (like with a `delay` or `yield`) to allow times for the buffers to clear because we are not in the `loop` context.

Fundamentally the "proper" way to handle this is to rebuild the UI protocol so that the client Javascript actually acknowledges messages and so the ESP and client _coordinate_ rather than the ESP just attempting to dump 45MB of JSON over the wires and hoping it all makes it. It isn't hard, just a little fiddly. I am working on a PR which should fix this soon.

For the moment, just send the CSS update a half second after `UI.begin` and it should be fine.

**Comment by thomastech** [comment]
Thanks for the detailed explanations, really helps me understand the world of CSS. I'm an old firmware programmer and web UI's are magic to me.

> For the moment, just send the CSS update a half second after UI.begin and it should be fine.

I was hoping that you meant the delay would fix the width issue. But the fussy width still defaults to the big size (until I change the entry).  If you have a workaround then I would be grateful.

I'm also finding weird behavior with the Slider controls. Just as a sanity check, can you please test a slider and see if it acts strange to you? What I'm finding is that moving the Slider sends the correct message, but upon releasing the Slider it reverts back to the default position. I suppose I can fix that with value updates, but seems odd I'd have to do that.

- Thomas

**Comment by iangray001** [comment]
Web UIs are held together with duct tape and dreams. I hate web technologies :)

Sliders seem to work fine for me (Firefox/Safari/Mobile Safari). I'll have a more thorough look, although it might take a few days because I'm deep in fixing the underlying transport issue first.

**Comment by thomastech** [comment]
@iangray001 

> Web UIs are held together with duct tape and dreams. 

Good thing I'm a dreamer. And along with duct tape,  I also have an old spool of rusty bailing wire.

> Sliders seem to work fine for me (Firefox/Safari/Mobile Safari).

Thanks for confirming that you observe working Sliders. That gave me a place to start.

Through a process of elimination and blind luck, I found out what causes the Sliders to revert back to their default setting. This may be a bug or something I'm doing wrong. 

Before I spam the Github Issues with a new ticket, I'll post a snippet of my code here (maybe someone will see a mistake). I don't show the volAdjust() function, but have confirmed it is not involved.

```
    uint16_t homeTab  = ESPUI.addControl(ControlType::Tab, "HOME", "Home");
    uint16_t ctrlTab  = ESPUI.addControl(ControlType::Tab, "CONTROL", "Control");

    homeRdsTmrID = ESPUI.addControl(ControlType::Label, "RDS MESSAGE TIMER", "DUMMY TEXT", ControlColor::Peterriver, homeTab);

    ctrlVolumeID = ESPUI.addControl(ControlType::Slider, "USB VOLUME", "20", ControlColor::Wetasphalt, ctrlTab, &volAdjust);
    ESPUI.addControl(ControlType::Min, "MIN", "0",   ControlColor::None, ctrlVolumeID);
    ESPUI.addControl(ControlType::Max, "MAX", "30",  ControlColor::None, ctrlVolumeID);
```

Sliders work correctly if Value updates are not made to the homeRdsTmrID Label control. But at one second intervals I have a task that does this:
```
     tempStr = String(timeCnt) + " Secs";
     ESPUI.updateControlValue(homeRdsTmrID, tempStr);
```

Updating the Label Value causes the Slider knob to move back to the default position ("20").

I recognize that you aren't customer support and I appreciate your advice. And maybe @s00500 can advise as well. If there are no obvious mistakes (or it needs further investigation) then I will report this as a separate issue. 

- Thomas

**Comment by iangray001** [comment]
Thomas, you're correct. Sliders work fine for me..._until they are programmatically updated_. 

I've fixed the underlying transport issue that was inherent to this library previously (not merged quite yet, more testing required), so now UIs of practically infinite size will work fine. This means I have been able to build a monster test UI that "does everything, to everything" and yes I can see that updating sliders is flaky. I'll get on it.

**Comment by iangray001** [comment]
Fixed this now. It will make its way into the branch I linked above.

**Comment by thomastech** [comment]
Fantastic. Will be nice to have Sliders that stay where I put them.

- Thomas
-
